### PR TITLE
ENH PHP 8.1 compatibility

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-framework/issues/10250

### CODE REVIEW

File: code/LeftAndMain.php
Comment (GuySartorelli) [inline code/LeftAndMain.php:1768]: By this point if `$lockModules[$moduleName]` was not null (i.e. `!isset()`) we've already returned out of the method. No need for null coalescing here.

File: code/MemberImportForm.php
Comment (GuySartorelli) [inline code/MemberImportForm.php:40]: This is ultimately the same as the `GroupCsvBulkLoader` import spec as mentioned above.
No need for null coalescing here.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Found a lot of situations where null coalescing just isn't necessary - and I stopped because there were just too many of them. We've discussed this a bit in person but for the sake of having this recorded somewhere can you please respond with your thinking on this? It's definitely overkill the way this is being done so far and it'll be good to have it recorded that this is a conscious decision.

**Comment by emteknetnz** [comment]
> It's definitely overkill the way this is being done so far and it'll be good to have it recorded that this is a conscious decision.

Yes agree it's overkill, though I'm not sure how else we can do this.

The code writing simply looks to see if a variable is being passed in as a parameter, rather than a scalar or a nested native function call with a single return type.

Ideally the code writer would have more smarts about it where it could judge if modification being made to a variable could lead it to being in a nullable state.

However even if we did have a method that could trace back changes made to variables, we're still stuck with the fact that the vast majority of Silverstripe methods have dynamic return types, and the docblock return types are very often wrong, so you cannot rely on them.  Since variables are largely a result of calls to Silverstripe methods, we won't very far with an automated cleanup.

At a future date, once Silverstripe is fully strongly typed, I would like to do an automated cleanup run that does trace what type variables, and then based on that remove the redundant null coalescing statements

File: code/CMSBatchActionHandler.php
Comment (emteknetnz) [reply code/CMSBatchActionHandler.php:287]: Correct fix here would be some defensive programming like adding a new line about this such as `$ids = is_array($ids) ? $ids : [];`

However since this hasn't come up in any unit test failures, I think we should just leave it as is.  (There will be thousands of other examples similar to this)

**Comment by GuySartorelli** [comment]
In general I agree - doing this in an automated fashion we're either going to go overboard and miss a few cases, or be light-handed and miss _heaps_ of cases. But on the otherhand doing this all by hand (like I was for the review) is a long laborious process with little payoff. Ultimately having some null coalescing operators where they aren't strictly needed isn't going to hurt anybody, and like you say it'll be much easier to go back and clean that up in a future major release when things are strongly typed.

I'll give Max and other core committers a chance to review this discussion before moving on but I think this is ultimately aan acceptable approach.

**Comment by maxime-rainville** [comment]
I think Steve and I had a similar discussion in a previous Zoom call. Generally, my mind is in the same place yours are. There's not really a way to support PHP8.1 without warning and without having a big automating process that will type cast everything.

To a degree, this will bite us in the ass later once we strongly type everything ... and will potentially neglect some of the benefit of strongly typing because it will obfuscate logic errors. 

In the end, we have to begrudgingly accept the automated approach and as a short term necessary evil.

**Review by GuySartorelli — APPROVED** [review]
As per the above conversation, the purposes of this PR is to avoid passing null (specifically null) from built-in functions that as of php 8.1 will throw deprecation warnings if null is passed in. The approach is necessarily heavy-handed, and while there are many situations where it isn't needed, it would be prohibitively laborious to find all of those situations by hand.

With that context, this looks good to me.

File: code/ModelAdmin.php
Comment (GuySartorelli) [inline code/ModelAdmin.php:623]: An error is emitted whether this is an empty string or null.
> Uncaught ValueError: Path cannot be empty

Is a warning emitted _as well_ as the error if this is an empty string?

**Review by GuySartorelli — COMMENTED** [review]
Just a question here before I merge.

File: code/ModelAdmin.php
Comment (emteknetnz) [reply code/ModelAdmin.php:623]: both, php 8.1
```
php -a
php > file_get_contents(null);
PHP Deprecated:  file_get_contents(): Passing null to parameter #1 ($filename) of type string is deprecated in php shell code on line 1

Deprecated: file_get_contents(): Passing null to parameter #1 ($filename) of type string is deprecated in php shell code on line 1
PHP Warning:  Uncaught ValueError: Path cannot be empty in php shell code:1
Stack trace:
#0 php shell code(1): file_get_contents()
#1 {main}
  thrown in php shell code on line 1

Warning: Uncaught ValueError: Path cannot be empty in php shell code:1
Stack trace:
#0 php shell code(1): file_get_contents()
#1 {main}
  thrown in php shell code on line 1

```

**Pull request merged**
